akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/scaladsl/ClusterSharding.scala

```diff
@@ -290,9 +286,23 @@ final class Entity[M, E] private[akka] (
 }
 
 /**
- * Parameter to [[Entity.apply]]
+ * Parameter to `createBehavior` function in [[Entity.apply]].
+ *
+ * Cluster Sharding is often used together with [[akka.persistence.typed.scaladsl.EventSourcedBehavior]]
+ * for the entities. To make that combination convenient to use the `EntityContext` has a `persistenceIdProposal`
+ * which can be used as the [[PersistenceId]] of the `EventSourcedBehavior`. It is constructed from the `name`
+ * of the [[EntityTypeKey]] and the `entityId` by concatenating them with `|` separator, or the separator that was
+ * defined by [[EntityTypeKey.withEntityIdSeparator]]. It's not mandatory to use exactly this `PersistenceId`.
+ * If `EventSourcedBehavior` isn't used the `persistenceIdProposal` can be ignored.
+ *
+ * @param entityId the business domain identifier of the entity
+ * @param persistenceIdProposal suggestion of [[PersistenceId]] to be used with an
+ *                              [[akka.persistence.typed.scaladsl.EventSourcedBehavior]]
  */
-final class EntityContext(val entityId: String, val shard: ActorRef[ClusterSharding.ShardCommand])
+final class EntityContext(
+    val entityId: String,
+    val persistenceIdProposal: PersistenceId,
```

Comment:
@renatocaval is this name ok? I was thinking that it's good to start with `persisteceId` but still give a hint about that it's optional

Reply:
I'm fine with just `persistenceId`. I don't think 'proposal' is making it clearer. 

I find it strange to name it 'proposal' it gives the impression that's something that can be rejected. Or something that you have to branch (if/else) to make a decision.

Alternatively, we could give the responsibility to the `PersistenceId` class instead of using the `EntityTypeKey`. Correct me if I'm wrong, but there is no hard requirement that the name of the entity needs to be the same as the type hint of the persistence id. Then we have the recommended way of doing it, ie: `PersistenceId(typeHint = "Account", id = ctx.entityId)` and the legacy way `PersistenceId(ctx.entityId)`. 

And if I'm not using sharding, I get also exposed, via API, to the recommended way.

I would expect a user to define the entity key as `EntityTypeKey[AccountCommand]("Account")`, but it should also be fine if defined as `EntityTypeKey[AccountCommand]("account-model")`

Maybe we are just creating that problem because we are trying to reuse the name.

Reply:
I just read your email where you point a previous decision to make the releation with sharding and persistence more clear. I'm aware that this is the intent and I don't want to re-open this discussion. So, please ignore my proposal to move it to the `PersistenceId`. 

Unless you find it really cool. ;-)

Reply:
Personally, I like the suggestion from @renatocaval to have a type parameter in `PersistenceId.apply`. It's true that not only cluster sharding use cases need to disambiguate the ID.

Reply:
If we want to get rid of the persistence stuff in sharding api I think your suggestion makes sense. We would document it as the recommended way to construct a PersistenceId when used together with Sharding would be:
```
PersistenceId(typeHint = Account.TypeKey.name, id = entityContext.entityId)
```

but one can use whatever other format

I have been thinking about if the EntityTypeKey should be included in the EntityContext. Wouldn't hurt I guess, even though it's typically already in scope from where `init` is callled.

Then it could look like this:
```
PersistenceId(typeHint = entityContext.entityTypeKey.name, id = entityContext.entityId)
```




we would also have to have an optional `separator` parameter.

Reply:
> I have been thinking about if the EntityTypeKey should be included in the EntityContext. Wouldn't hurt I guess, even though it's typically already in scope from where `init` is callled.

IMHO, it makes sense to have the `EntityTypeKey` in the `EntityContext` so the function `EntityContext => Behavior[M]` has all the information it needs. Of course users can close over the `EntityTypeKey` when creating the `EntityContext => Behavior[M]` but I found it lacking that `EntityContext` didn't give me all the _context_ual information.

> Then it could look like this:
> 
> ```
> typeHint = entityContext.entityTypeKey.name
> id = entityContext.entityId
> ```

👆🏼 This looks great.

Reply:
(let me put it on a separate comment to keep things separate)

> `PersistenceId(typeHint = ..., id = entityContext.entityId)`

I also like that, putting sharding aside, `PersistenceId` now has an extra field.

Reply:
Summing my two comments above: I think having access on the `EntityTypeKey` from the `EntityContext` and adding the `typeHint` on the `PersistenceId` provide a flexible enough API for users.

When we're sharding projection workers in Lagom, the `EntityTypeKey.name` has no relation with the `PersistenceId.typeHint`. Instead, `EntityTypeKey.name` is the prefix used by the tagger.

So adding this flexibility in `EntityTypeKey` and `PersistenceId` will allow a very simple integration with projections code.